### PR TITLE
Add scalability good practices doc

### PR DESCRIPTION
**Pull request by gmarek**

### CODE REVIEW

File: contributors/devel/scalability-good-practices.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+If you're changing some big and important dependency, there's no way of knowing what performance impact it'll have. The only thing we can do is to actually run test and check.
+Where to look to get data?
+If you want to check the impact of your changes there are a number of places you can look at:
+Density/Load tests output quite a lot of data either to test logs, or files inside 'ReportDir' - both of them include API call latencies and Density also includes pod e2e startup latency information,
+for resource usage you can either use monitoring tools (heapster + graphana, but at the time of writing it stops working around 100 Nodes), or just plain old top on master (which scales as much as you want:),
+more data is available on `/metrics` endpoint on all our components (e.g. one for the API server contains API call latencies),
+to profile a component is to create an ssh tunnel to the machine running it, and run `go tool pprof localhost:<your_tunnel_port>` locally
+
+## Summary
+Summing it up, when writing code you should:
```
Comment (thockin) [inline contributors/devel/scalability-good-practices.md]: Did you mean this to be a bulleted list?

File: contributors/devel/scalability-good-practices.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+ 
+4. From time to time you need to add a very complex logic to the system, one computation of which on big clusters may take a lot of time. You need to be sure that you won't add this logic on a critical path of any of standard flows.
+ 
+5. When updating big dependencies (especially etcd or golang version) you need to run large scale test to verify that bump doesn't break Kubernetes performance.
+
+## What does it mean to "break scalability"?
+By "breaking scalability" we mean causing performance SLO violation in one of our performance tests. Performance SLOs for Kubernetes are <sup>[2](#2)</sup>:
+- 99th percentile of API call latencies <= 1s
+- 99th percentile of e2e Pod startup, excluding image pulling, latencies <= 5s
+Tests that we run are Density and Load, we invite everyone interested in details to read the code.
```
Comment (wojtek-t) [inline contributors/devel/scalability-good-practices.md]: Break the line here (see how it looks in md file).

File: contributors/devel/scalability-good-practices.md (hunk 1 above)
Comment (wojtek-t) [inline contributors/devel/scalability-good-practices.md]: +1

File: contributors/devel/scalability-good-practices.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+ 
+What we did to mitigate this issue was, in a way, reimplementing Informers using GETs instead of WATCHes. Current solution consists of a `Secret` cache shared between all `Pod`s. When a `Pod` wants to check if `Secret` changed to looks into the cache. If the `Secret` stored in the cache is too old, cache issues a GET request to API server to get current value. As `Pods` within a single `Namespace` share the `Secret` for `ServiceAccount` it means that Kubelet will need to refresh the `Secret` only once a while per `Namespace`, not per `Pod`, as it was before. This of course is a stopgap, not a final solution, which is currently (as of early May 2017) being designed as a "Bulk Watch".
```
Comment (wojtek-t) [inline contributors/devel/scalability-good-practices.md]: You can link the design at the end: https://github.com/kubernetes/community/pull/443

File: contributors/devel/scalability-good-practices.md (hunk 1 above)
Comment (gmarek) [inline contributors/devel/scalability-good-practices.md]: Done.

**Comment by gmarek** [comment]
Done. @wojtek-t PTAL

**Pull request merged**